### ISSUE TITLE
[Investigation] How to get metrics in a pod efficiently

### ISSUE THREAD

**Issue by KONY128** [tass-io/scheduler#3]
After the investigation on metrics in a pod, I confirm that without the query of control plane on Kubernetes, metrics are impossible to get in a K-native way

The metrics are collected by Kubelet from the queries on CRI. And then send status update to the control plan through ApiServer 

The idea of getting metrics inside a pod is barely promoted. 

The good news is, I think the kernel binaries like ``free`` maybe can get some performance metrics, I'm trying to do this

Any progress will be posted here

**Comment by KONY128** [comment]
After investigation and some actual practices in a pod:

Metrics can be got inside a pod by some naive queries into the Linux system like ``/proc/stat`` and ``/proc/meminfo``.

Using ``pprof`` library is the best choice to get the **instant metrics** of CPU and Memory usage.

The callings of Linux binary like ``top`` and ``free`` are high-latency and resource-inefficient, so they should not be considered to.

The usage of [gopsutil](https://github.com/shirou/gopsutil) is out of our consideration because the CPU profile is too inaccurate.

If we send query to CRI, Kubelet or even ApiServer, the cost of it will be much more the than binary execution in a pod. So method like it is not suggested.

**Comment by linxuyalun** [comment]
> After investigation and some actual practices in a pod:
> 
> Metrics can be got inside a pod by some naive queries into the Linux system like `/proc/stat` and `/proc/meminfo`.
> 
> Using `pprof` library is the best choice to get the **instant metrics** of CPU and Memory usage.
> 
> The callings of Linux binary like `top` and `free` are high-latency and resource-inefficient, so they should not be considered to.
> 
> The usage of [gopsutil](https://github.com/shirou/gopsutil) is out of our consideration because the CPU profile is too inaccurate.
> 
> If we send query to CRI, Kubelet or even ApiServer, the cost of it will be much more the than binary execution in a pod. So method like it is not suggested.

I have read an [article](https://medium.com/google-cloud/continuous-profiling-of-go-programs-96d4416af77b) about `pprof` from raykll in medium, and I think it's a good ref.

TLDR:  "_We target an additional 5% overhead for CPU and heap allocation profiling. The collection is happening for 10 seconds for every minute from a single instance._"
